compiler/ast/checked_ast.nim

```diff
@@ -0,0 +1,112 @@
+## The ``CheckedAst`` type is implemented here, along with its accessor
+## routines and routines for creating and manipulating ``CheckedAst``
+## instances.
+##
+## ``CheckedAst`` is a wrapper around ``PNode``, and is intended to make
+## propagation less manual, by keeping track of whether there is an error
+## *somewhere* in the managed AST.
+##
+## The idea is to use ``CheckedAst`` as the return type of all semantic
+## analysis procedures, eventually removing the need for wrapper errors
+## (``adWrappedError``).
+
+import
+  compiler/ast/[
+    ast,
+    ast_types,
+    errorhandling,
+    lineinfos
+  ],
+  compiler/front/[
+    options
+  ]
+
+type
+  CheckedAst* = object
+    ## Represents either in-progress or fully semanticized AST.
+    n: PNode       ## stores the AST. Never nil
+    hasError: bool ## tracks whether `n` contains an ``nkError`` node
+                   ## somewhere
+
+proc prepare*(x: var CheckedAst, n: PNode) =
```

Comment:
```suggestion
proc initForMappingChildren*(x: var CheckedAst, n: PNode) =
```

The name is _huge_, but I'm hoping it can help inspire some ideas and we can land on something better.

Additionally, I think this should ensure that `n` has children, otherwise, I don't think it really makes sense that we `shallowCopy`; feels like the purpose is off.